File: 3sum-closest/3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        def m(a, b):
            if a > b: return a-b
            else: return b-a
        
        nums.sort()
        ret_sum=0
        min_diff = float('inf')
        
        for i in range(len(nums)-2):
            a = i + 1
            b = len(nums) - 1
            
            while a < b:
                sum = nums[i] + nums[a] + nums[b]
                if m(sum,target) < min_diff:
                    min_diff = m(sum,target)
                    ret_sum = sum
                if sum > target:
                    b -= 1
                elif sum < target:
                    a += 1
                else:
                    break
        
        return ret_sum

File: 3sum-closest/3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int minSum = nums[0] + nums[1] + nums[2];
        for(int i=0; i<nums.length-2; i++) {
            int lo = i+1, hi = nums.length-1;
            while(lo < hi) {
                int sum = nums[i] + nums[lo] + nums[hi];
                if (Math.abs(target-sum) < Math.abs(target-minSum)) {
                    minSum = sum;
                }
                if (sum == target) {
                    minSum = target;
                    break;
                } else if (sum < target) {
                    lo++;
                } else if (sum > target) {
                    hi--;
                }
            }
        }
        return minSum;
    }
}